### PR TITLE
Proxy example

### PR DESCRIPTION
**Pull request by legionth**
Currently based on #17 

Little example that shows that you can create a HTTP proxy server with this project.

### CODE REVIEW

**Comment by clue** [comment]
Nice! :+1: I'll review this more thoroughly once #17 is in.

**Comment by legionth** [comment]
Ping @clue

File: examples/proxy.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+require __DIR__ . '/../vendor/autoload.php';
+
+$loop = React\EventLoop\Factory::create();
+
+$callback = function(RequestInterface $request) use ($loop){
+    return new Promise(function ($resolve, $reject) use ($request, $loop) {
+        $connector = new TcpConnector($loop);
+        $resolverFactory = new React\Dns\Resolver\Factory();
+        $resolver = $resolverFactory->create('8.8.8.8', $loop);
+        $dnsConnector = new DnsConnector($connector, $resolver);
```
Comment (clue) [inline examples/proxy.php:20]: Should probably only be injected once?

File: examples/proxy.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+
+        $host = $request->getHeader('Host')[0];
```
Comment (clue) [inline examples/proxy.php:23]: Tested across all versions? What if the host header is unset?

File: examples/proxy.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        $hostArray = parse_url($host);
```
Comment (clue) [inline examples/proxy.php:24]: What if this fails?

File: examples/proxy.php (hunk 2 above)
Comment (clue) [reply examples/proxy.php:23]: What kind of request is being handled here? How is this verified at all?

File: examples/proxy.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        if (isset($hostArray['host'])) {
+            $host = $hostArray['host'];
+        }
+
+        // Create to given Host:Port
+        $dnsConnector->create($host, $port)->then(function ($stream) use ($request, $resolve){
+            $body = $request->getBody();
+            $body->on('data', function ($chunk) use ($resolve, $stream) {
+                $stream->write($chunk);
+            });
```
Comment (clue) [inline examples/proxy.php:40]: backpressure?

File: examples/proxy.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        if (isset($hostArray['port'])) {
+            $port = $hostArray['port'];
+        }
+
+        if (isset($hostArray['host'])) {
+            $host = $hostArray['host'];
+        }
+
+        // Create to given Host:Port
+        $dnsConnector->create($host, $port)->then(function ($stream) use ($request, $resolve){
```
Comment (clue) [inline examples/proxy.php:36]: What if this fails?

File: examples/proxy.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            $body = $request->getBody();
+            $body->on('data', function ($chunk) use ($resolve, $stream) {
+                $stream->write($chunk);
+            });
+
+            $responseBuffer = '';
+            $stream->on('data', function ($data) use (&$responseBuffer){
+                $responseBuffer .= $data;
+            });
```
Comment (clue) [inline examples/proxy.php:45]: Why does this need to buffer? What if the response is huge?

**Review by clue — CHANGES_REQUESTED** [review]
Plenty of open issues here afaict, does it really make sense to address these here?

If we can provide a small example here, then I'm all for adding this here. Otherwise, *perhaps* move this elsewhere?

**Comment by legionth** [comment]
This example would be to large to be a simple example. Will be handled in an seperated repo.

**Pull request closed without merging**